### PR TITLE
feat!: Update OTel Hook to conform to official conventions

### PR DESCRIPTION
**Pull request by james-milligan**
Signed-off-by: James Milligan <james@omnant.co.uk>

## This PR

- Updates OTel Hook to conform to official conventions

### Related Issues

https://github.com/open-feature/go-sdk-contrib/issues/49

### CODE REVIEW

File: hooks/open-telemetry/pkg/otel.go [hunk 1]
```diff
@@ -2,165 +2,60 @@ package otel
...
-}
-
-// WithContext sets the parent context used when new spans are formed. When the parent context is closed so are all spans.
-// if no parent context is provided context.Background() is used internally
-func (h *Hook) WithContext(ctx context.Context) {
-	h.ctx = ctx
-}
-
-// Before creates the flag evaluations open-telemetry span and sets the FlagKey and ProviderName attributes
+// Before method is unused for this hook
```
Comment (skyerus) [inline hooks/open-telemetry/pkg/otel.go:29]: Could include `UnimplementedHook` from the go-sdk in your Hook struct to avoid implementing empty funcs

Comment (james-milligan) [reply hooks/open-telemetry/pkg/otel.go]: updated

File: hooks/open-telemetry/pkg/otel.go [hunk 2]
```diff
@@ -2,165 +2,53 @@ package otel
...
-	AttributeFlagKey = "feature_flag.flag_key"
-	// AttributeProviderName trace attribute name for the provider used for evaluation
-	AttributeProviderName = "feature_flag.provider_name"
-	// AttributeEvaluatedVariant trace attribute name for the variant returned from the flag evaluation
-	AttributeEvaluatedVariant = "feature_flag.evaluated_variant"
-	// AttributeEvaluatedValue trace attribute name for the value returned from the flag evaluation
-	AttributeEvaluatedValue = "feature_flag.evaluated_value"
-	// AttributeTraceName the name of the trace exported by the hook
-	AttributeTraceName = "github.com/open-feature/go-sdk-contrib/hooks/open-telemetry/pkg"
+	EventName                 = "feature_flag"
```
Comment (skyerus) [inline hooks/open-telemetry/pkg/otel.go:12]: Do these need to be exported?

Comment (james-milligan) [reply hooks/open-telemetry/pkg/otel.go:12]: Ive kept this exported so that the test package can be kept separate, whilst checking for the correct value

**Review by skyerus — APPROVED** [review]

File: hooks/open-telemetry/pkg/otel.go [hunk 3]
```diff
@@ -2,165 +2,53 @@ package otel
...
-	}
-	mw.ss.span.SetAttributes(
-		attribute.String(AttributeEvaluatedValue, value),
+	span := trace.SpanFromContext(h.ctx)
+	span.SetAttributes(
+		attribute.String(EventPropertyFlagKey, hookContext.FlagKey()),
+		attribute.String(EventPropertyProviderName, hookContext.ProviderMetadata().Name),
+		attribute.String(EventPropertyVariant, flagEvaluationDetails.Variant),
 	)
+	span.AddEvent(EventName)
```
Comment (beeme1mr) [inline hooks/open-telemetry/pkg/otel.go:42]: It looks like `AddEvent` accepts attributes. Here's an example: https://github.com/MrAlias/opentelemetry-go/blob/46582921368bcafa47ae4d5b17cca9b9e20e0a5a/example/basic/main.go#L91

File: hooks/open-telemetry/pkg/otel.go [hunk 4]
```diff
@@ -2,165 +2,53 @@ package otel
...
-			return err
-		}
-		value = string(val)
-	default:
-		return fmt.Errorf("unknown data type received: %d", flagEvaluationDetails.FlagType)
-	}
-	mw.ss.span.SetAttributes(
-		attribute.String(AttributeEvaluatedValue, value),
+	span := trace.SpanFromContext(h.ctx)
+	span.SetAttributes(
```
Comment (beeme1mr) [inline hooks/open-telemetry/pkg/otel.go:37]: This doesn't seem correct. The attributes should be associated with the span event not the span.

**Review by beeme1mr — CHANGES_REQUESTED** [review]
This doesn't appear to follow the OTel spec. The attributes must be on the event not the span itself.

**Comment by james-milligan** [comment]
> This doesn't appear to follow the OTel spec. The attributes must be on the event not the span itself.

Ive made the requested changes, the attributes are now associated with the event, rather than the span itself

File: hooks/open-telemetry/pkg/otel.go [hunk 5]
```diff
@@ -2,165 +2,52 @@ package otel
...
-		return errors.New("no span stored for provided hook context")
-	}
-	if flagEvaluationDetails.ResolutionDetail.Variant != "" {
-		mw.ss.span.SetAttributes(
-			attribute.String(AttributeEvaluatedVariant, flagEvaluationDetails.ResolutionDetail.Variant),
-		)
+// After sets the feature_flag event and associated attributes on the span stored in the context
+func (h *Hook) After(hookContext openfeature.HookContext, flagEvaluationDetails openfeature.InterfaceEvaluationDetails, hookHints openfeature.HookHints) error {
+	if h.ctx == nil {
+		// if no context is set trace.SpanFromContext will return a noop span
```
Comment (toddbaert) [inline hooks/open-telemetry/pkg/otel.go:33]: Is it better to simply return the no-op span then?

I'm trying to understand the benefits of handling this ourselves by returning `nul`, vs avoiding this check and returing a noop span.

File: hooks/open-telemetry/pkg/otel.go [hunk 6]
```diff
@@ -2,165 +2,52 @@ package otel
...
-type storedSpan struct {
-	cancel func()
-	span   trace.Span
+	ctx context.Context
+	openfeature.UnimplementedHook
 }
 
 // NewHook return a reference to a new instance of the OpenTelemetry Hook
-func NewHook() *Hook {
+func NewHook(ctx context.Context) *Hook {
```
Comment (toddbaert) [inline hooks/open-telemetry/pkg/otel.go:24]: Should we somehow warn of an invalid state if this constructor is not used and the context is unset?

Comment (james-milligan) [reply hooks/open-telemetry/pkg/otel.go]: Unexpected the hook struct so the use of `NewHook` is forced, past this point I think the log is unnecessary, as you would have to pass a nil context into the constructor as you can no longer bypass this by creating the hook directly with `hook := otelhook.Hook{}`

File: hooks/open-telemetry/pkg/otel.go (hunk 5 above)
Comment (james-milligan) [reply hooks/open-telemetry/pkg/otel.go]: ive removed the nil check and allowed for the loop hook to be used

**Review by beeme1mr — APPROVED** [review]
Approved by please fix the linting error.

**Pull request merged**